src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User actAs(X x, User superUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(superUser.getId());
+
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null ) {
+      throw new AuthenticationException("Super user doesn't exist.");
```

Comment:
`AuthenticationException` was created solely for the purpose of being thrown when a user is trying to do something without being logged in. If the front end makes a remote procedure call that throws an `AuthenticationException`, it will show the user the login screen. That makes sense because the user should have to log in before doing anything else in the system. However, being shown the login screen when you _are_ logged in doesn't make sense. We should use other exception types for those cases.

That's why in the previous PR I asked if most of these exceptions could be changed. Currently if you logged in as yourself and then tried to act as someone who doesn't exist, you'll be shown the login screen, which I don't think makes sense. I think we should throw a RuntimeException here instead to avoid showing the login screen to the user.

* AuthenticatedException means you weren't logged in when you tried to do something
* AuthorizationException means you were logged in but you didn't have permission to do something
* RuntimeException means something went wrong but we don't have a more specific exception to throw that would be more appropriate

Reply:
> If the front end makes a remote procedure call that throws an AuthenticationException, it will show the user the login screen.

That's not true. I still stand by what I said in the previous PR, the agent auth service is a form of authentication and should throw authentication exceptions where necessary.

Reply:
> That's not true.

I'm pretty sure it is unless I misunderstood the code here: https://github.com/foam-framework/foam2/blob/master/src/foam/box/SessionClientBox.js#L30 Isn't that doing what I said?

Reply:
Whoops, you're right. I was testing the junction authorization exception when checking to see if the authentication exception redirected user to login screen. I'll make the changes where necessary.